ER図・README作成

95yuki0305 opened the pull request:
# What
ER図・READMEを作成

# Why
どんなデータベース設計をしていくのかの基盤を確認するため
ER図のURL(https://gyazo.com/75e8726ea9255d2a85c59768d7f1e189)

95yuki0305 committed:
ER図・README作成

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| user            | references | null: false, foreign_key: true |
+
+### Association
+belongs_to :user
```
cardsテーブルならびに他テーブルのcardsテーブルとのアソシエーションを削除しましょう！

理由
今回必須機能においては購入の度にカード情報を入力するため、カード情報を保存する必要はない為です。

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| family name      | string  | null: false |
+| first name       | string  | null: false |
+| family name ruby | string  | null: false |
+| first name ruby  | string  | null: false |
```
カラム名はスネークケースの単数形がルールとなっております。
よって複数単語が有る場合は、user_nameというように_(アンダーバー)で繋げましょう！

***他のカラム名も同様、複数単語が有る場合、_で繋げるようにしましょう。***

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| ---------------- | ------- | ----------- |
+| nickname         | string  | null: false |
+| email            | string  | null: false |
+| password         | string  | null: false |
```
9行目のpasswordカラムの表記を修正しましょう。

理由
gemのdeviseを使用する場合、DBに保存されるのは、deviseが予め用意している `encrypted_password`というカラムになるからです。
新規登録の際に`password`というキーに対して入力された値は暗号化されて`encrypted_password`というカラムに保存されます。

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| Column           | Type    | Options     |
+| ---------------- | ------- | ----------- |
+| nickname         | string  | null: false |
+| email            | string  | null: false |
```
```ruby
| email               | string | null: false |
```
下記のユーザー管理機能の実装条件に合うように、Optionsに記述を追記しましょう。

```ruby
メールアドレスが一意性であること（unique: true）
```

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| first name       | string  | null: false |
+| family name ruby | string  | null: false |
+| first name ruby  | string  | null: false |
+| birthday         | integer | null: false |
```
```ruby
| birthday         | integer | null: false |
```
タイプを修正しましょう。

理由
今回の実装では、誕生日カラムを一つにまとめて頂き、タイプは「date型」で設計した方が実装が簡単になる為です。
date型にすると、年、月、日を1つで指定して取得できます。

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+
+### Association
+- has_many :products
+- has_one :buyer
```
```ruby
 has_one :buyer
```

ユーザーは１個しか商品が買えない仕様にはなっていないかと思われます。
ご確認後修正をしましょう。

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| status           | string     | null: false                    |
+| shipping charges | string     | null: false                    |
+| shipping area    | string     | null: false                    |
+| days to ship     | string     | null: false                    |
```
```ruby
| category         | string     | null: false                    |
| status           | string     | null: false                    |
| shipping charges | string     | null: false                    |
| shipping area    | string     | null: false                    |
| days to ship     | string     | null: false                    |
```
27〜31行目のプルダウンで実装するデータたちについてカラム名とタイプを修正しましょう。

理由
trelloの商品出品に【カテゴリー選択など、基本的に変更されないデータを取り扱う際は、以下のカリキュラムを参考にして、ActiveHashを使用すること。】とありますので、商品出品の際に必要な「 カテゴリー」,「商品の状態」,「 配送料の負担」, 「発送元の地域」, 「発送までの日数」合計5つのカラムは、active_hashで実装しましょう。
active_hashで実装するカラム名は`末尾に「_id」`をつけ、`「単数系カラム名_id」`にしましょう。
そしてタイプはすべて`integer型`にしましょう。

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| user             | references | null: false, foreign_key: true |
+
+### Association
+belongs_to :user
```
```ruby
belongs_to :user
```

こちら、購入履歴テーブルのアソシエーションも必要かと思われます。
どの商品が売られたか分かるようにするためです。

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+
+| Column        | Type       | Options                        |
+| ------------- | ---------- | ------------------------------ |
+| postal code   | integer    | null: false                    |
```
郵便番号はstring型で実装しましょう。
理由は数字の間にハイフン（-）を考慮するためです。（例：111-1111）

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| Column        | Type       | Options                        |
+| ------------- | ---------- | ------------------------------ |
+| postal code   | integer    | null: false                    |
+| prefectures   | string     | null: false                    |
```
こちら商品テーブルにいるActivehashと同様、カラムとタイプを修正しましょう。
またActivehashは使い回すことが可能ですので、商品テーブルにいる都道府県と同じ名前にしましょう。

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+### Association
+belongs_to :user
+
+## buyers テーブル
```
住所情報が保存されるテーブルと、購入履歴を保存するテーブルを分けましょう。

【住所テーブル】
商品購入時に配送先住所を入力する際に必要な、`市町村、番地、建物番号、郵便番号、電話番号の情報`が保存されるテーブルが今回の実装では必要となります。
アドレスは購入履歴テーブルとひも付けが欲しい為、アソシエーションも設定しましょう。

【購入履歴履歴テーブル】
誰が何を買ったか分かるよう、ユーザーと商品の外部キーをセットしたテーブルを作成しましょう。
また商品と、ユーザーと、アドレスとひも付けが欲しい為、アソシエーションも設定しましょう。

div-reviewer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| municipality  | string     | null: false                    |
+| address       | string     | null: false                    |
+| building name | string     |                                |
+| phone number  | integer    | null: false                    |
```
電話番号はstring型で実装しましょう。

理由
電話番号は0からスタートするためです。

補足
integer型の場合、仮に「01234」を保存しようとすると先頭の0が読み込まれず「1234」に変換されてしまいます。数値(integer)ではなく文字列(string)として扱うことで先頭に0が来た場合でも「01234」と保存することが可能になります。

div-reviewer reviewed (commented):
最終課題、初レビュー依頼ありがとうございます！

合計【12点】コメントさせていただきました。
細かい内容は1つにまとめさせて頂いている箇所もありますので、お手数ですがご確認お願いします。

また、コードを修正していただいたあとは、ER図の方も記載内容が変わるかと思いますので、こちらも修正をお願い致します。

ご不明点や疑問点ございましたら、お手数ですがコネクトメンターにぜひご相談ください。

どうぞよろしくおねがいします！

95yuki0305 committed:
cardsテーブルの削除及び、他テーブルとのアソシエーションを削除

95yuki0305 committed:
カラム名の複数単語を_で繋げる

95yuki0305 committed:
カラム名のpasswordをencrypted_passwordに変更

95yuki0305 committed:
メールアドレスを一意性に変更

95yuki0305 committed:
birthdayのtypeをdate型に変更

95yuki0305 committed:
has_many :buyersに変更

95yuki0305 committed:
カラム名に_idをつけ、タイプをinteger型に変更

95yuki0305 committed:
Revert "has_many :buyersに変更"

This reverts commit cf202a3b0a1592aaa921a1db9aff92139df0e57f.

95yuki0305 committed:
has_many :buyersに変更

95yuki0305 committed:
購入履歴テーブルとのアソシエーションを定義

95yuki0305 committed:
郵便番号のカラムの型をstring型に変更

95yuki0305 committed:
prefectures_idに変更

・usersテーブルのアソシエーションを変更
・productsテーブルのアソシエーションを変更
・buyersテーブルをaddressテーブルに変更

95yuki0305 committed:
住所テーブルと購入履歴テーブルの作成

95yuki0305 committed:
電話番号のカラムの型をstring型に変更

95yuki0305 committed:
ER図　修正

95yuki0305 committed:
Revert "ER図　修正"

This reverts commit a4aa8f5fc8d17b71b64962bcb2c2fdb0ebdd39f7.

95yuki0305 committed:
ER図　修正

https://gyazo.com/e4ba4278bd4a3c32d4a5f7712c1407b9

shodai-ito commented on README.md:
```diff
@@ -1,24 +1,65 @@
 
-* Ruby version
+### Association
+- has_many :products, through:product_users
```
`has_many :products, through:product_users`
product_userテーブルは中間デーブルではないため、throughオプションは不要です！

shodai-ito commented on README.md:
```diff
@@ -1,24 +1,65 @@
-* Ruby version
+### Association
+- has_many :products, through:product_users
+- has_many :address
```
`has_many :address`
必須機能の段階ではユーザーと住所はアソシエーションを組む必要はありません！
こちらのアソシエーションは削除しましょう！

shodai-ito replied:
product_userとのアソシエーションを記述しましょう！

shodai-ito commented on README.md:
```diff
@@ -1,24 +1,65 @@
-* Database creation
+### Association
+belongs_to :user
+has_many :users, through:product_users
```
`has_many :users, through:product_users`
userとはbelongs_toでつながっているのでこちらの記述は不要です！

また、購入履歴のテーブルとのアソシエーションを記述しましょう！

shodai-ito commented on README.md:
```diff
@@ -1,24 +1,65 @@
-* Deployment instructions
+- belongs_to :product
+- belongs_to :user
+- belongs_to :address
```
`belongs_to :address`
アソシエーションはhas_oneにしましょう！

shodai-ito commented on README.md:
```diff
@@ -1,24 +1,65 @@
+| house_number     | string     | null: false                    |
+| building_name    | string     |                                |
+| phone_number     | string     | null: false                    |
+| user             | references | null: false, foreign_key: true |
```
`| user             | references | null: false, foreign_key: true |`
userとのアソシエーションは不要ですので削除しましょう！

また、products_userとアソシエーション を組みますので、products_userの外部キーカラムを作成しましょう！

shodai-ito commented on README.md:
```diff
@@ -1,24 +1,65 @@
+| user             | references | null: false, foreign_key: true |
+
+### Association
+belongs_to :user
+has_many :product_users
```
userとのアソシエーションは削除しましょう！
また、product_userとのアソシエーションはbelongs_toに変更しましょう！

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました！
コメントしましたので、ご確認をお願いいたします！

その後、再度レビュー依頼フォームのご提出をよろしくお願いいたします。

95yuki0305 committed:
usersテーブルのアソシエーションthroughオプションを削除

95yuki0305 committed:
usersテーブルのアソシエーションでhas_many :addressを削除

95yuki0305 committed:
productsテーブルのアソシエーションを修正

95yuki0305 committed:
アソシエーションをhas_oneに変更

95yuki0305 committed:
addressテーブルのアソシエーションを修正

95yuki0305 committed:
ER図　修正

https://gyazo.com/d9cd157b2c07b1976212545087dc2b56

togo-mori commented on README.md:
```diff
@@ -1,24 +1,64 @@
 
-* Ruby version
+### Association
+- has_many :product_user
```
```
- has_many :product_user
```
has_manyの場合は複数形で記述するため、 `:product_user` → `:product_users` 
としましょう！

togo-mori commented on README.md:
```diff
@@ -1,24 +1,64 @@
-* Database creation
+### Association
+belongs_to :user
+has_many :product_user
```
```
has_many :product_user
```
productsテーブルからproduct_usersテーブルへのアソシエーションは
`has_one` で設定しましょう！

理由
今回のフリマでは同じ商品は1回しか購入できない仕様にするため。

補足
一部ECサイトでは商品に在庫があり複数回購入可能なものもございますが
今回のフリマでは上記のような仕様で作成していただきたく存じます。

togo-mori commented on README.md:
```diff
@@ -1,24 +1,64 @@
+| category_id         | integer     | null: false                    |
+| status_id           | integer     | null: false                    |
+| shipping_charges_id | integer     | null: false                    |
+| prefectures_id      | integer     | null: false                    |
```
```
| prefectures_id      | integer     | null: false                    |
```
細かい点ですがこちらのカラム名は `prefecture_id` と「単数形_id」の形式にしましょう。

**理由**
 ActiveHashを用いたカラムの命名規則によるものです。 ActiveHashを用いて保存するカラムのカラム名はモデル名+_idとなります。 モデル名はrailsでは単数形で定義する決まりになっているため、カラム名が単数形+_idの形式でない場合 モデル名の命名規則と合わなくなってしまいます。

togo-mori commented on README.md:
```diff
@@ -1,24 +1,64 @@
+| Column           | Type       | Options                        |
+| ---------------- | ---------- | ------------------------------ |
+| postal_code      | string     | null: false                    |
+| prefectures_id   | integer    | null: false                    |
```
```
| prefectures_id   | integer    | null: false                    |

```
こちらも下記コメントと同様に `prefecture_id` と単数形にしましょう！
https://github.com/95yuki0305/furima-35586/pull/1#discussion_r630666788

togo-mori reviewed (commented):
レビュー依頼ありがとうございます！
いくつかコメントしましたので確認・修正お願いいたします。

修正後再度フォームにてレビュー依頼お願いいたします。

95yuki0305 committed:
複数形の記述に変更

95yuki0305 committed:
productsテーブルのアソシエーションを修正

has_many :product_userをhas_oneに変更

95yuki0305 committed:
単数形_idに変更.ER図修正

prefectures_id → prefecture_id
Gyazo(https://gyazo.com/caada79df96f3fb9f0f428724ca41721)

Hide-Morishita commented:
レビュー依頼ありがとうございます。
修正点確認させていただきました。

ER図の確認をさせていただきたいため、
お手数ですが、**READMEの内容を反映させた、ER図の提出**をお願いいたします。

95yuki0305 commented:
# What
ER図です。お願いします。(https://gyazo.com/caada79df96f3fb9f0f428724ca41721)

# Why
READMEの内容を反映したER図の確認をしてもらうため。

95yuki0305 commented:
# What
ER図です。お願いします。(https://gyazo.com/caada79df96f3fb9f0f428724ca41721)

# Why
READMEの内容を反映したER図の確認をしてもらうため。

d-kazu4k reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます。実装お疲れさまでした！

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き頑張っていきましょう！
